### ISSUE TITLE
Generate URL (without hyperlink) for CURL.

### ISSUE THREAD

**Issue by XtremeOwnageDotCom** [mkdocs/mkdocs#3853]
I have a use-case- I am building documentation, which contains sections with code blocks. TO, give a very brief example-

-----------------------------------------------------------------------------

[Download](my-script.sh)

To run directly from bash, run this:

``` bash
curl https://thiswebsite/path/blah/my-script.sh
```

------------------------------------------------------------------------------

So, making a hyperlink can automatically link to a file, such as...

```
[myscript](my-script.sh)
```

But, how can I generate this link, and use it inside of a code-block? (Without- it being an actual hyper-link, Just the bare URL)

I have mkdocs-macros. I am all ears for a solution to this problem.

The closest solution I currently have, is a custom jinja2 filter, which uses page.canonical url and removes the last path-segment, to where I then manually append the file name.

------------------------------------------------------------------------------

In short, I want to be able to do something like one of these:

```
curl [my-script](my-script.sh){nohtml}
curl {{ file_url('my-script.sh') }}
```

and have this generated:

```
curl https://somedomain.com/docs/path/my-script.sh
```

And, ideally, I will get a warning if the URL is not valid, when building.

**Comment by XtremeOwnageDotCom** [comment]
@squidfunk Didn't figure this one belonged in mkdocs-material- but, if you have any ideas, I'm all ears!

**Comment by XtremeOwnageDotCom** [comment]
I did, find a solution using mkdocs-macros.

``` python
    @env.macro
    def fileuri(filename):
        # Check if the "files" object is defined
        files = env.variables.get('files')
        page = env.variables.get('page')
        if not files:
            raise ValueError("Error: 'files' object is not defined in the environment.")
        if not page:
            raise ValueError("Error: 'page' object is not defined in the environment.")
       
        # Get the directory of the current page
        current_dir = os.path.dirname(env.page.file.abs_src_path)

        # First, search for the file relative to the current directory
        for file in files:
            if os.path.basename(file.src_path) == filename and os.path.dirname(file.src_path) == current_dir:
                return file.dest_uri
        
        # If not found, fall back to searching for the file name and extension across all files
        for file in files:
            if os.path.basename(file.src_path) == filename:
                return file.dest_uri
        
        # If file is not found, raise an error
        raise ValueError(f"Error: File '{filename}' not found in the 'files' object.")
```

**Comment by squidfunk** [comment]
Thanks for the mention – I believe it does not belong into Material for MkDocs, but rather into the extension that processes the code blocks, which is likely the Highlight extension and/or Pygments.

**Comment by XtremeOwnageDotCom** [comment]
> Thanks for the mention – I believe it does not belong into Material for MkDocs, but rather into the extension that processes the code blocks, which is likely the Highlight extension and/or Pygments.

I'd say it belonged a level above that- in the case for example you wanted to give the URL for a script hosted in the docs repo directly on a page... for whatever reason.

The simplified use-case, is inserting the direct URI for a given file hosted within the docs repo.

For my case though, this macro should do the trick.

![image](https://github.com/user-attachments/assets/f1e7b27c-28c6-4029-bf0b-ea985924fdf0)
